## ** 실습 과제2 <11주차 과제 - 100점>**

### **(1) 각각의 문제 바로 아래에 답을 작성 후, Bigdata_project2.ipynb 파일을 한림SmartLEAD에 제출할 것**
### **(2) 제출 데드라인 : 2022.11.07(월요일) ~ 2022.11.13(일요일) 23:59 까지**
### **(3) 부정행위 적발 시, 원본(보여준 사람)과 복사본(베낀 사람) 모두 0점 처리**
### **(4) !!해당 문제의 코드와 출력 결과를 반드시 기록할 것(둘 중 하나라도 없으면 해당 문제 0점처리)!!**

- - -



### **1. 다음의 의료비 예측 데이터 세트 테이터를 사용하여 Charges를 반응변수(종속변수)로 하는 ``다중회귀분석(Multiple Linear Regression)``을 진행하시오.**

``* 다중회귀분석 50점``

**<의료비 예측 데이터 세트>**

  * age: age of primary beneficiary
  * sex: insurance contractor gender, female, male
  * bmi: Body mass index, providing an understanding of body, weights that are relatively high or low relative to height,
objective index of body weight (kg / m ^ 2) using the ratio of height to weight, ideally 18.5 to 24.9
  * children: Number of children covered by health insurance / Number of dependents
  * smoker: Smoking
  * region: the beneficiary's residential area in the US, northeast, southeast, southwest, northwest.
  * charges: Individual medical costs billed by health insurance

<br>

#### **다음의 조건을 만족해야 함**

```
## 조건
1.	Charges를 반응변수로 사용하고 그 외 나머지를 설명변수로 사용.
2.	createDataPartition을 이용하여 Train/Test set 분리 후 회귀모델 생성. 
3.	Summary함수를 사용하여 회귀분석 결과 해석을 작성.
4.	분석 결과에 기반하여Feature Selection(변수 선택) 및 새로운 회귀모델 생성.
5.	Test set을 이용한 Predict(예측)결과 비교.
```



### **1. 다중회귀분석 (50점)**

* 아래의 코드 작성란에 코드 작성후 결과 출력하세요.

In [18]:
#다중회귀분석 시 필요한 패키지 설치
install.packages(c("tidyverse", "data.table", "caret", "e1071", "Hmisc"))
library(tidyverse)
library(data.table)
options(repr.plot.width=10, repr.plot.height=10)

Installing packages into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘listenv’, ‘parallelly’, ‘future’, ‘globals’, ‘future.apply’, ‘numDeriv’, ‘progressr’, ‘SQUAREM’, ‘lava’, ‘prodlim’, ‘deldir’, ‘RcppEigen’, ‘iterators’, ‘Rcpp’, ‘clock’, ‘gower’, ‘hardhat’, ‘ipred’, ‘timeDate’, ‘png’, ‘jpeg’, ‘interp’, ‘checkmate’, ‘htmlwidgets’, ‘foreach’, ‘ModelMetrics’, ‘plyr’, ‘pROC’, ‘recipes’, ‘reshape2’, ‘proxy’, ‘Formula’, ‘latticeExtra’, ‘gridExtra’, ‘htmlTable’, ‘viridis’


Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──
✔ ggplot2 3.3.6      ✔ purrr   0.3.5 
✔ tibble  3.1.7      ✔ dplyr   1.0.10
✔ tidyr   1.2.1      ✔ stringr 1.4.1 
✔ readr   2.1.3      ✔ forcats 0.5.2 
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks

In [92]:
#사용할 데이터셋을 불러옵니다.
df <- read.csv("/content/sample_data/insurance.csv")

#데이터셋을 확인합니다.
head(df, n = 10)

#구조를 확인합니다.
str(df)

age sex    bmi    children smoker region    charges  
1  19  female 27.900 0        yes    southwest 16884.924
2  18  male   33.770 1        no     southeast  1725.552
3  28  male   33.000 3        no     southeast  4449.462
4  33  male   22.705 0        no     northwest 21984.471
5  32  male   28.880 0        no     northwest  3866.855
6  31  female 25.740 0        no     southeast  3756.622
7  46  female 33.440 1        no     southeast  8240.590
8  37  female 27.740 3        no     northwest  7281.506
9  37  male   29.830 2        no     northeast  6406.411
10 60  female 25.840 0        no     northwest 28923.137

'data.frame':	1337 obs. of  7 variables:
 $ age     : int  19 18 28 33 32 31 46 37 37 60 ...
 $ sex     : chr  "female" "male" "male" "male" ...
 $ bmi     : num  27.9 33.8 33 22.7 28.9 ...
 $ children: int  0 1 3 0 0 0 1 3 2 0 ...
 $ smoker  : chr  "yes" "no" "no" "no" ...
 $ region  : chr  "southwest" "southeast" "southeast" "northwest" ...
 $ charges : num  16885 1726 4449 21984 3867 ...


In [107]:
#다중회귀분석을 수행합니다.
#조건 1. Charges를 반응변수로 사용, 나머지는 설명변수로 사용
model_1 = lm(charges ~ age+sex+bmi+children+smoker+region, data = df)
model_1


Call:
lm(formula = charges ~ age + sex + bmi + children + smoker + 
    region, data = df)

Coefficients:
    (Intercept)              age          sexmale              bmi  
       -11936.6            256.8           -129.5            339.3  
       children        smokeryes  regionnorthwest  regionsoutheast  
          474.8          23847.3           -349.2          -1035.3  
regionsouthwest  
         -960.1  


In [108]:
summary(model_1)


Call:
lm(formula = charges ~ age + sex + bmi + children + smoker + 
    region, data = df)

Residuals:
     Min       1Q   Median       3Q      Max 
-11305.1  -2850.3   -979.9   1395.0  29992.8 

Coefficients:
                 Estimate Std. Error t value Pr(>|t|)    
(Intercept)     -11936.56     988.23 -12.079  < 2e-16 ***
age                256.76      11.91  21.555  < 2e-16 ***
sexmale           -129.48     333.20  -0.389 0.697630    
bmi                339.25      28.61  11.857  < 2e-16 ***
children           474.82     137.90   3.443 0.000593 ***
smokeryes        23847.33     413.35  57.693  < 2e-16 ***
regionnorthwest   -349.23     476.82  -0.732 0.464053    
regionsoutheast  -1035.27     478.87  -2.162 0.030804 *  
regionsouthwest   -960.08     478.11  -2.008 0.044836 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 6064 on 1328 degrees of freedom
Multiple R-squared:  0.7507,	Adjusted R-squared:  0.7492 
F-statistic:   500 on 8 an

In [109]:
#조건2 : createDataPartition을 이용하여 Train/Test set 분리 후 회귀모델 생성.
#createDataPartition을 이용하여 Train/Test set을 분리합니다.
index <- caret::createDataPartition(y = df$charges, p = 0.8, list = FALSE)
train <- df[index, ]
test <- df[-index, ]

train %>% show()
test %>% show()

     age    sex    bmi children smoker    region   charges
1     19 female 27.900        0    yes southwest 16884.924
2     18   male 33.770        1     no southeast  1725.552
3     28   male 33.000        3     no southeast  4449.462
4     33   male 22.705        0     no northwest 21984.471
5     32   male 28.880        0     no northwest  3866.855
6     31 female 25.740        0     no southeast  3756.622
7     46 female 33.440        1     no southeast  8240.590
8     37 female 27.740        3     no northwest  7281.506
9     37   male 29.830        2     no northeast  6406.411
10    60 female 25.840        0     no northwest 28923.137
11    25   male 26.220        0     no northeast  2721.321
12    62 female 26.290        0    yes southeast 27808.725
13    23   male 34.400        0     no southwest  1826.843
14    56 female 39.820        0     no southeast 11090.718
15    27   male 42.130        0    yes southeast 39611.758
16    19   male 24.600        1     no southwest  1837.2

In [110]:
#회귀모델 생성
model_2 = lm(charges ~ age+sex+bmi+children+smoker+region, data = train)
model_2


Call:
lm(formula = charges ~ age + sex + bmi + children + smoker + 
    region, data = train)

Coefficients:
    (Intercept)              age          sexmale              bmi  
       -12038.6            264.7           -125.1            333.1  
       children        smokeryes  regionnorthwest  regionsoutheast  
          325.8          24017.8           -256.6           -814.6  
regionsouthwest  
         -821.1  


- 회귀분석 결과
1. 회귀식 도출
* ```
b0 = -12038.6, b1 = 264.7, b2 = -125.1, b3 = 333.1, b4 = 325.8, b5 = 24017.8, b6 = -256.6, b7 = -814.6, b8 =  -821.1으로부터
charges =-12038.6 + 264.7 * age + -125.1 * sex + ...
``` 
2. age, sex, bmi, children, smoker, region이 각각 1 증가할 때, charges는 264.7, -125.1, 333.1, 325.8, 24017.8이 증가

In [111]:
#조건 3 : Summary 함수를 사용하여 회귀분석 결과 해석을 작성.
summary(model_2)


Call:
lm(formula = charges ~ age + sex + bmi + children + smoker + 
    region, data = train)

Residuals:
     Min       1Q   Median       3Q      Max 
-11699.8  -2887.5   -960.3   1366.9  29529.3 

Coefficients:
                 Estimate Std. Error t value Pr(>|t|)    
(Intercept)     -12038.59    1132.29 -10.632   <2e-16 ***
age                264.72      13.44  19.701   <2e-16 ***
sexmale           -125.08     373.36  -0.335   0.7377    
bmi                333.06      32.45  10.262   <2e-16 ***
children           325.76     155.42   2.096   0.0363 *  
smokeryes        24017.84     464.69  51.686   <2e-16 ***
regionnorthwest   -256.55     539.42  -0.476   0.6344    
regionsoutheast   -814.65     538.68  -1.512   0.1308    
regionsouthwest   -821.07     535.73  -1.533   0.1257    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 6093 on 1063 degrees of freedom
Multiple R-squared:  0.752,	Adjusted R-squared:  0.7501 
F-statistic: 402.9 on 8 

<결과의 해석>
```
1. 제일 밑에 F-statistic의 P-value 값은 2.2e -16로 0.05보다 작기에 이 회귀식은 회귀분석 모델 전체에 대해 통계적으로 의미가 있다고 볼 수 있다.

2. 중간의 Coefficients : 에는 y절편 값(Intercept) 및 변수들의 p-value값이 나와있다.
sex, region 변수의 경우 0.05보다 크기에 charges를 설명하는데 유의하지 않다.

3. 밑 부분의 Adjusted R-squared 값은 0.7501로, 75%만큼의 설명력을 가진다고 판단할 수 있다.

``` 

In [112]:
# 조건4. 분석 결과에 기반하여 Feature Selection(변수 선택) 및 새로운 회귀모델 생성
#유의하지 않은 변수를 누락하고 회귀모형을 새롭게 정의
model_3 = lm(charges ~ age+bmi+children+smoker, data = train)
model_3


Call:
lm(formula = charges ~ age + bmi + children + smoker, data = train)

Coefficients:
(Intercept)          age          bmi     children    smokeryes  
   -12220.5        265.9        319.8        324.2      23980.5  


In [113]:
summary(model_3)


Call:
lm(formula = charges ~ age + bmi + children + smoker, data = train)

Residuals:
     Min       1Q   Median       3Q      Max 
-12216.5  -2919.8   -933.6   1342.4  29157.2 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) -12220.46    1074.10  -11.38   <2e-16 ***
age            265.93      13.41   19.83   <2e-16 ***
bmi            319.79      30.94   10.34   <2e-16 ***
children       324.20     155.15    2.09   0.0369 *  
smokeryes    23980.55     461.91   51.92   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 6092 on 1067 degrees of freedom
Multiple R-squared:  0.7512,	Adjusted R-squared:  0.7502 
F-statistic: 805.2 on 4 and 1067 DF,  p-value: < 2.2e-16


In [114]:
#조건5 : Test set을 이용한 Predict(예측)결과 비교.
#test 데이터 확인
test %>% show()

     age    sex    bmi children smoker    region   charges
21    60 female 36.005        0     no northeast 13228.847
25    37   male 28.025        2     no northwest  6203.902
29    23   male 17.385        1     no northwest  2775.192
30    31   male 36.300        2    yes southwest 38711.000
35    28   male 36.400        1    yes southwest 51194.559
39    35   male 36.670        1    yes northeast 39774.276
40    60   male 39.900        0    yes southwest 48173.361
44    37 female 30.800        2     no southeast  6313.759
47    18 female 38.665        2     no northeast  3393.356
49    60 female 24.530        0     no southeast 12629.897
52    21 female 33.630        2     no northwest  3579.829
57    58 female 31.825        2     no northeast 13607.369
68    40   male 26.315        1     no northwest  6389.378
71    27 female 24.750        0    yes southeast 16577.780
72    31   male 28.500        5     no northeast  6799.458
73    53 female 28.100        3     no southwest 11741.7

In [116]:
#테스트 데이터를 활용하여 예측 시작
model_3_r <- predict(model_3, test)
test <- test %>% bind_cols(tibble(model_3_predict = model_3_r))
test

New names:
• `model_3_predict` -> `model_3_predict...8`
• `model_3_predict` -> `model_3_predict...9`


age sex    bmi    children smoker region    charges   model_3_predict...8
1   60  female 36.005 0        no     northeast 13228.847 15249.1989         
2   37  male   28.025 2        no     northwest  6203.902  7229.3639         
3   23  male   17.385 1        no     northwest  2775.192  -220.3843         
4   31  male   36.300 2        yes    southwest 38711.000 32260.6492         
5   28  male   36.400 1        yes    southwest 51194.559 31170.6568         
6   35  male   36.670 1        yes    northeast 39774.276 33118.4767         
7   60  male   39.900 0        yes    southwest 48173.361 40475.3403         
8   37  female 30.800 2        no     southeast  6313.759  8116.7902         
9   18  female 38.665 2        no     northeast  3393.356  5579.3872         
10  60  female 24.530 0        no     southeast 12629.897 11579.5711         
11  21  female 33.630 2        no     northwest  3579.829  4767.0035         
12  58  female 31.825 2        no     northeast 13607.369 14029.0055         
13  40  male   26.315 1        no     northwest  6389.378  7156.0964         
14  27  female 24.750 0        yes    southeast 16577.780 26854.9438         
15  31  male   28.500 5        no     northeast  6799.458  6758.3041         
16  53  female 28.100 3        no     southwest 11741.726 11832.3464         
17  29  female 29.590 1        no     southeast  3947.413  5278.2432         
18  45  female 38.285 0        no     northeast  7935.291 11989.4509         
19  57  female 31.160 0        yes    northwest 43578.939 36882.5718         
20  56  male   19.950 0        yes    northeast 22412.648 33031.7642         
21  30  male   25.460 0        no     northeast  3645.089  3899.2257         
22  20  male   28.025 1        yes    northwest 17560.380 26364.9874         
23  63  male   35.090 0        yes    southeast 47055.532 39734.9100         
24  37  male   30.800 0        no     southwest  4646.759  7468.3975         
25  32  female 17.765 2        yes    northwest 32734.186 26599.2061         
26  53  female 35.900 2        no     southwest 11163.568 14002.5376         
27  29  male   29.735 2        no     northwest 18157.876  5648.8095         
28  46  male   30.495 3        yes    northwest 40720.551 34717.3222         
29  51  female 37.730 1        no     southeast  9877.608 13731.7127         
30  48  male   29.700 0        no     southeast  7789.635 10041.8011         
⋮   ⋮   ⋮      ⋮      ⋮        ⋮      ⋮         ⋮         ⋮                  
236 48  female 36.575 0        no     northwest  8671.191 12240.3798         
237 19  female 27.930 3        no     northwest 18838.704  2736.5280         
238 59  male   37.100 1        no     southwest 12347.172 15657.6437         
239 52  female 33.300 2        no     southwest 10806.839 12905.1500         
240 18  male   39.140 0        no     northeast 12890.058  5082.8963         
241 38  female 30.210 3        no     northwest  7537.164  8518.2336         
242 40  male   24.970 2        no     southeast  6593.508  7050.1708         
243 58  male   30.305 0        no     northeast 11938.256 12894.5270         
244 26  male   31.065 0        no     northwest  2699.568  4627.9665         
245 37  male   22.705 3        no     northeast  6985.507  5852.2601         
246 24  male   29.830 0        yes    northeast 18648.422 27681.7183         
247 32  female 20.520 0        no     northeast  4544.235  2851.2972         
248 28  male   37.100 1        no     southwest  3277.161  7413.9655         
249 64  male   23.760 0        yes    southeast 26926.514 36377.5774         
250 25  female 34.485 0        no     northwest  3021.809  5455.7344         
251 32  female 29.735 0        no     northwest  4357.044  5798.1921         
252 39  male   29.925 1        yes    northeast 22462.044 32025.1716         
253 18  female 21.660 0        yes    northeast 14283.459 23473.4567         
254 61  male   36.300 1        yes    southwest 47403.880 39914.2060         
255 46  male   25.745 3   


### **2. 다음의 뇌졸증 예측 데이터 세트 테이터를 사용하여 stroke를 반응변수(종속변수)로 하는 ``로지스틱 회귀분석(Logistic Regression)``을 진행하시오.**

``* 로지스틱 회귀분석 50점``

**<뇌졸증 예측 데이터 세트>**

  * id: unique identifier
  * gender: "Male", "Female" or "Other"
  * age: age of the patient
  * hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension
  * heart_disease: 0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
  * ever_married: "No" or "Yes"
  * work_type: "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
  * Residence_type: "Rural" or "Urban"
  * avg_glucose_level: average glucose level in blood
  * bmi: body mass index
  * smoking_status: "formerly smoked", "never smoked", "smokes" or "Unknown"*
  * stroke: 1 if the patient had a stroke or 0 if not

<br>

#### **다음의 조건을 만족해야 함**

```
## 조건
1. Stroke Prediction 데이터를 로드 후 다음과 같은 전처리 과정을 거친다.
   - 결측값 제거(bmi 컬럼값이 N/A로 작성되어 있는 값을 결측값으로 지정함)
   - 결측값 제거 후 bmi 데이터타입 변경 
   - Stroke 값이 0인 데이터를 랜덤으로 209개 추출
   - 추출된 데이터와 stroke 값이 1인 데이터 병합
2. 병합한 Stroke Prediction 데이터를 caret 패키지의 createDataPartition을 이용하여 train / test 으로 분리할 것 
3. Train dataset을 사용하여glm 함수를 사용하여 모델을 생성하고 반응변수는 stroke로 설정하고, 설명변수는 stroke를 제외한 모든 것을 사용한다. 
   - step, update을 생성한 모델을 새롭게 정의할 것. 
   - 생성한 모델과 새롭게 정의한 모델의 summary결과를 확인한다.
4. Test set을 사용해서 모델을 예측한다. 예측 결과와 원본 데이터와 비교하여 모델의 적합성에 대하여 기술한다. 

```



In [26]:
#조건1 : Stroke Prediction 데이터를 로드 후 다음과 같은 전처리 과정을 거친다.
#사용할 데이터셋을 불러옵니다.
df <- read.csv("/content/sample_data/Stroke Prediction Dataset.csv")

#데이터셋을 확인합니다.
head(df, n = 10)

#구조를 확인합니다.
str(df)

id    gender age hypertension heart_disease ever_married work_type
1   9046 1      67  0            1             1            4        
2  51676 2      61  0            0             1            5        
3  31112 1      80  0            1             1            4        
4  60182 2      49  0            0             1            4        
5   1665 2      79  1            0             1            5        
6  56669 1      81  0            0             1            4        
7  53882 1      74  1            1             1            4        
8  10434 2      69  0            0             2            4        
9  27419 2      59  0            0             1            4        
10 60491 2      78  0            0             1            4        
   Residence_type avg_glucose_level bmi  smoking_status stroke
1  1              228.69            36.6 1              1     
2  2              202.21            N/A  2              1     
3  2              105.92            32.5 2              1     
4  1              171.23            34.4 3              1     
5  2              174.12            24   2              1     
6  1              186.21            29   1              1     
7  2               70.09            27.4 2              1     
8  1               94.39            22.8 2              1     
9  2               76.15            N/A  0              1     
10 1               58.57            24.2 0              1

'data.frame':	5110 obs. of  12 variables:
 $ id               : int  9046 51676 31112 60182 1665 56669 53882 10434 27419 60491 ...
 $ gender           : int  1 2 1 2 2 1 1 2 2 2 ...
 $ age              : num  67 61 80 49 79 81 74 69 59 78 ...
 $ hypertension     : int  0 0 0 0 1 0 1 0 0 0 ...
 $ heart_disease    : int  1 0 1 0 0 0 1 0 0 0 ...
 $ ever_married     : int  1 1 1 1 1 1 1 2 1 1 ...
 $ work_type        : int  4 5 4 4 5 4 4 4 4 4 ...
 $ Residence_type   : int  1 2 2 1 2 1 2 1 2 1 ...
 $ avg_glucose_level: num  229 202 106 171 174 ...
 $ bmi              : chr  "36.6" "N/A" "32.5" "34.4" ...
 $ smoking_status   : int  1 2 2 3 2 1 2 2 0 0 ...
 $ stroke           : int  1 1 1 1 1 1 1 1 1 1 ...


In [27]:
summary(df)

       id            gender           age         hypertension    
 Min.   :   67   Min.   :1.000   Min.   : 0.08   Min.   :0.00000  
 1st Qu.:17741   1st Qu.:1.000   1st Qu.:25.00   1st Qu.:0.00000  
 Median :36932   Median :2.000   Median :45.00   Median :0.00000  
 Mean   :36518   Mean   :1.586   Mean   :43.23   Mean   :0.09746  
 3rd Qu.:54682   3rd Qu.:2.000   3rd Qu.:61.00   3rd Qu.:0.00000  
 Max.   :72940   Max.   :3.000   Max.   :82.00   Max.   :1.00000  
 heart_disease      ever_married     work_type     Residence_type 
 Min.   :0.00000   Min.   :1.000   Min.   :1.000   Min.   :1.000  
 1st Qu.:0.00000   1st Qu.:1.000   1st Qu.:2.000   1st Qu.:1.000  
 Median :0.00000   Median :1.000   Median :4.000   Median :1.000  
 Mean   :0.05401   Mean   :1.344   Mean   :3.495   Mean   :1.492  
 3rd Qu.:0.00000   3rd Qu.:2.000   3rd Qu.:4.000   3rd Qu.:2.000  
 Max.   :1.00000   Max.   :2.000   Max.   :5.000   Max.   :2.000  
 avg_glucose_level     bmi            smoking_status      stro

In [28]:
#조건 1-1 : 결측값 제거(bmi 컬럼값이 N/A로 작성되어 있는 값을 결측값으로 지정함)
df[df$bmi == "N/A",]=NA

#N/A값이 NA로 바꼈는지 확인
head(df, n= 10)

id    gender age hypertension heart_disease ever_married work_type
1   9046  1     67   0            1             1            4       
2     NA NA     NA  NA           NA            NA           NA       
3  31112  1     80   0            1             1            4       
4  60182  2     49   0            0             1            4       
5   1665  2     79   1            0             1            5       
6  56669  1     81   0            0             1            4       
7  53882  1     74   1            1             1            4       
8  10434  2     69   0            0             2            4       
9     NA NA     NA  NA           NA            NA           NA       
10 60491  2     78   0            0             1            4       
   Residence_type avg_glucose_level bmi  smoking_status stroke
1   1             228.69            36.6  1              1    
2  NA                 NA            NA   NA             NA    
3   2             105.92            32.5  2              1    
4   1             171.23            34.4  3              1    
5   2             174.12            24    2              1    
6   1             186.21            29    1              1    
7   2              70.09            27.4  2              1    
8   1              94.39            22.8  2              1    
9  NA                 NA            NA   NA             NA    
10  1              58.57            24.2  0              1

In [29]:
#결측값 제거
df = na.omit(df)

In [30]:
#결측값 제거 됐는지 확인
colSums(is.na(df))

id            gender               age      hypertension 
                0                 0                 0                 0 
    heart_disease      ever_married         work_type    Residence_type 
                0                 0                 0                 0 
avg_glucose_level               bmi    smoking_status            stroke 
                0                 0                 0                 0

In [31]:
#조건 1-2 : 결측값 제거 후 bmi 데이터타입 변경
df$bmi <- as.numeric(df$bmi)

#bmi의 데이터타입이 numeric으로 변경 됐는지 확인
str(df$bmi)

 num [1:4909] 36.6 32.5 34.4 24 29 27.4 22.8 24.2 29.7 36.8 ...


In [32]:
#조건 1-3 : Stroke 값이 0인 데이터를 랜덤으로 209개 추출

#stroke 값이 0인 것을 새로운 변수에 정의합니다.
df_stroke_zero <- df[df$stroke == 0, ]

df_stroke_zero

id    gender age   hypertension heart_disease ever_married work_type
250  30669 1       3    0            0             2            1        
251  30468 1      58    1            0             1            4        
252  16523 2       8    0            0             2            4        
253  56543 2      70    0            0             1            4        
254  46136 1      14    0            0             2            3        
255  32257 2      47    0            0             1            4        
256  52800 2      52    0            0             1            4        
257  41413 2      75    0            1             1            5        
258  15266 2      32    0            0             1            4        
259  28674 2      74    1            0             1            5        
260  10460 2      79    0            0             1            2        
261  64908 1      79    0            1             1            4        
262  63884 2      37    0            0             1            4        
263  37893 2      37    0            0             1            4        
264  67855 2      40    0            0             1            4        
265  25774 1      35    0            0             2            4        
266  19584 2      20    0            0             2            4        
267  24447 2      42    0            0             1            4        
268  49589 2      44    0            0             1            2        
269  17986 2      79    0            1             1            5        
270  29217 2      65    1            0             1            4        
271  72911 2      57    1            0             1            4        
272  47175 2      49    0            0             1            4        
273   4057 1      71    0            0             1            4        
274  48588 2      59    0            0             1            4        
275  70336 2      25    0            0             1            4        
276  66767 2      67    0            0             1            2        
277  45801 2      38    0            0             2            4        
278  36275 2      54    0            0             1            4        
279  11577 2      70    0            0             1            5        
⋮    ⋮     ⋮      ⋮     ⋮            ⋮             ⋮            ⋮        
5078  8203 1      17.00 0            0             2            4        
5079 27446 2       8.00 0            0             2            1        
5080 42709 1       1.72 0            0             2            1        
5081 22691 2      29.00 0            0             1            5        
5082 37680 1      55.00 0            0             1            2        
5083 24552 2      44.00 0            0             1            4        
5084 72914 2      19.00 0            0             2            4        
5085 29540 1      67.00 0            0             1            4        
5086 53525 2      72.00 0            0             1            4        
5087 65411 2      51.00 0            0             1            4        
5088 26214 2      63.00 0            0             1            5        
5089 22190 2      64.00 1            0             1            5        
5090 56714 2       0.72 0            0             2            1        
5091  4211 1      26.00 0            0             2            2        
5092  6369 1      59.00 1            0             1            4        
5093 56799 1      76.00 0            0             1            2        
5095 28048 1      13.00 0            0             2            1        
5096 68598 1       1.08 0            0             2            1        
5097 41512 1      57.00 0            0             1            2        
5098 64520 1      68.00 0            0             1            5        
5099   579 1       9.00 0            0             2            1        
5101 68398 1      82.00 1            0             1            5        
5102 3690

In [33]:
#sample함수를 사용하여 랜덤으로 209개를 추출합니다.
df_stroke_r <- df_stroke_zero[sample(nrow(df_stroke_zero), 209), ]

df_stroke_r

id    gender age hypertension heart_disease ever_married work_type
2582 29872 2      35  0            0             1            4        
2479 65643 2       7  0            0             2            1        
4502 24246 1       7  0            0             2            1        
2283 55566 2      34  0            0             1            4        
3925 39958 1      18  0            0             2            4        
3140 39984 2      42  0            0             1            2        
968   5694 1      21  0            0             2            4        
2884  4793 2      60  1            0             1            5        
397   2374 1      60  1            0             1            4        
4436 54117 1       7  0            0             2            1        
1664 20460 2      62  0            0             1            4        
913  11745 2      29  0            0             1            4        
4801 40850 2      74  0            0             1            2        
3771 67438 2      60  0            0             1            2        
809  21397 2      40  0            0             1            2        
1008 10762 2      41  0            0             1            4        
4420 31703 1      66  0            0             1            5        
2691 15070 1      76  0            1             1            4        
466    621 1      69  0            0             1            4        
1715 26893 1       8  0            0             2            1        
3760 53552 2      62  0            0             1            4        
2716 60003 1      81  0            0             1            2        
2724 31360 2      31  0            0             2            4        
4596 15061 1      40  0            0             1            2        
992  66494 1      48  0            0             1            4        
3411 28315 1      38  0            0             1            4        
2515 70625 1      18  0            0             2            4        
1714  4099 2      21  0            0             2            4        
5073 71957 2      35  0            0             1            4        
390  37090 2      70  0            0             1            4        
⋮    ⋮     ⋮      ⋮   ⋮            ⋮             ⋮            ⋮        
4734  7529 1      67  0            0             1            4        
2105 66680 2      49  1            0             1            4        
2114 35402 1      14  0            0             2            4        
3271 72082 2      45  0            0             1            5        
600  64174 2      59  1            0             1            4        
635   2532 1      28  0            0             2            4        
3656 57602 1       6  0            0             2            1        
4661 36793 2      38  1            0             1            4        
3541   187 2      20  0            0             2            4        
2524 30953 1      75  1            1             1            4        
3104 23171 1      66  0            0             1            4        
3613 50785 1      17  0            0             2            4        
1870  7885 2      23  0            0             2            4        
4496 30734 1      15  0            0             2            1        
4691 22590 1       5  0            0             2            1        
458  17718 2      33  1            0             1            4        
807  55529 1      39  0            0             1            4        
1268 43088 1      37  0            0             2            4        
3227 71414 2       2  0            0             2            1        
755  13367 2      35  0            0             1            4        
3834 51693 2      52  0            0             1            4        
1710 16316 1      35  0            0             1            4        
3893 47803 1      37  0            0             1            2        
2859 55462 1       6  0            0         

In [34]:
# 추출된 데이터와 stroke 값이 1인 데이터 병합

#stroke 값이 1인 것을 새로운 변수에 정의합니다.
df_stroke_one <- df[df$stroke == 1, ]

stroke_Prediction <- merge(df_stroke_one, df_stroke_r, all = T)

stroke_Prediction

id    gender age hypertension heart_disease ever_married work_type
1    187  2      20  0            0             2            4        
2    210  1      81  0            0             1            5        
3    354  2      65  0            0             1            4        
4    491  2      74  0            0             1            5        
5    621  1      69  0            0             1            4        
6    712  2      82  1            1             2            4        
7   1099  2      15  0            0             2            1        
8   1210  2      68  0            0             1            4        
9   1261  1      54  0            0             1            4        
10  1665  2      79  1            0             1            5        
11  1836  2      51  1            0             1            4        
12  2092  2      37  0            0             1            4        
13  2182  2      80  1            0             1            5        
14  2326  2      67  1            0             1            4        
15  2374  1      60  1            0             1            4        
16  2390  1      78  0            0             1            5        
17  2458  2      78  0            0             1            4        
18  2520  2      26  0            0             1            4        
19  2532  1      28  0            0             2            4        
20  2548  2      81  0            0             1            5        
21  2824  2      44  0            0             1            2        
22  3013  2      79  0            0             1            5        
23  3167  1      53  0            1             1            4        
24  3253  1      61  0            1             1            4        
25  3348  2      58  1            0             1            4        
26  3355  2      64  0            0             1            4        
27  3379  2      61  0            0             1            4        
28  3512  2      70  1            0             1            5        
29  3531  1      41  0            0             1            4        
30  4057  1      71  0            0             1            4        
⋮   ⋮     ⋮      ⋮   ⋮            ⋮             ⋮            ⋮        
389 68356 2      73  0            0             1            5        
390 68627 1      80  1            1             1            4        
391 68794 2      79  0            0             1            5        
392 68798 2      58  0            0             1            4        
393 69089 2      40  0            0             1            4        
394 69112 1      68  1            1             1            4        
395 69379 2      64  1            0             1            5        
396 69487 2      79  0            0             1            5        
397 69551 1      69  1            0             2            4        
398 69647 1      30  0            0             2            4        
399 69959 2      80  1            0             2            4        
400 70625 1      18  0            0             2            4        
401 70630 2      71  0            0             1            2        
402 70822 1      80  0            0             1            5        
403 70943 2      80  0            0             1            4        
404 71044 2       8  0            0             2            1        
405 71279 2      71  0            0             1            2        
406 71414 2       2  0            0             2            1        
407 71639 2      68  0            0             2            2        
408 71673 2      79  0            0             1            4        
409 71796 2      70  0            1             1            4        
410 71957 2      35  0            0             1            4        
411 72081 2      57  1            0             1            2        
412 72082 2      45  0            0             1            5        
413 72178 2       4  0      

### **2. 로지스틱 회귀분석 (50점)**

* 아래의 코드 작성란에 코드 작성후 결과 출력하세요.

In [35]:
#병합한 Stroke Prediction 데이터를 caret 패키지의 createDataPartition을 이용하여 train / test 으로 분리할 것 
index <- caret::createDataPartition(y = stroke_Prediction$stroke, p = 0.8, list = FALSE)
train <- stroke_Prediction[index, ]
test <- stroke_Prediction[-index, ]

train %>% show()
test %>% show()

       id gender   age hypertension heart_disease ever_married work_type
2     210      1 81.00            0             0            1         5
3     354      2 65.00            0             0            1         4
4     491      2 74.00            0             0            1         5
5     621      1 69.00            0             0            1         4
6     712      2 82.00            1             1            2         4
7    1099      2 15.00            0             0            2         1
8    1210      2 68.00            0             0            1         4
10   1665      2 79.00            1             0            1         5
11   1836      2 51.00            1             0            1         4
12   2092      2 37.00            0             0            1         4
13   2182      2 80.00            1             0            1         5
14   2326      2 67.00            1             0            1         4
15   2374      1 60.00            1             0  

In [36]:
str(train)

'data.frame':	336 obs. of  12 variables:
 $ id               : int  210 354 491 621 712 1099 1210 1665 1836 2092 ...
 $ gender           : int  1 2 2 1 2 2 2 2 2 2 ...
 $ age              : num  81 65 74 69 82 15 68 79 51 37 ...
 $ hypertension     : int  0 0 0 0 1 0 0 1 1 0 ...
 $ heart_disease    : int  0 0 0 0 1 0 0 0 0 0 ...
 $ ever_married     : int  1 1 1 1 2 2 1 1 1 1 ...
 $ work_type        : int  5 4 5 4 4 1 4 5 4 4 ...
 $ Residence_type   : int  2 1 1 2 2 2 2 2 1 2 ...
 $ avg_glucose_level: num  91.5 72.5 75 101.5 84 ...
 $ bmi              : num  31.4 28.9 26.6 26.8 26.5 22.2 39.3 24 28.4 27.5 ...
 $ smoking_status   : int  2 3 2 3 1 0 0 2 2 2 ...
 $ stroke           : int  1 0 1 0 1 0 1 1 1 0 ...


In [37]:
#Train dataset을 사용하여glm 함수를 사용하여 모델을 생성하고 반응변수는 stroke로 설정하고, 설명변수는 stroke를 제외한 모든 것을 사용한다. 

#glm(stroke ~., data = train) : stroke는 반응변수, .으로 모든 설명변수를 나타낸다.
# family = "binomial" :  종속 변수가 어떤 분포를 따르고 있는지 옵션을 주는 것
m <- glm(stroke ~., data = train, family = "binomial")
summary(m)


Call:
glm(formula = stroke ~ ., family = "binomial", data = train)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.09644  -0.76500   0.08044   0.77259   2.60667  

Coefficients:
                    Estimate Std. Error z value Pr(>|z|)    
(Intercept)       -4.608e+00  1.376e+00  -3.349 0.000811 ***
id                 1.174e-05  6.208e-06   1.891 0.058568 .  
gender            -1.108e-01  2.828e-01  -0.392 0.695184    
age                6.673e-02  9.827e-03   6.790 1.12e-11 ***
hypertension       6.270e-01  3.509e-01   1.787 0.073987 .  
heart_disease      3.603e-01  4.690e-01   0.768 0.442362    
ever_married      -4.196e-02  4.408e-01  -0.095 0.924175    
work_type          4.412e-03  1.502e-01   0.029 0.976572    
Residence_type    -3.413e-01  2.735e-01  -1.248 0.211963    
avg_glucose_level  2.166e-03  2.668e-03   0.812 0.416796    
bmi                1.896e-02  2.070e-02   0.916 0.359831    
smoking_status     7.257e-02  1.401e-01   0.518 0.604591    


In [38]:
#step, update을 생성한 모델을 새롭게 정의할 것. 

m2 <- step(m, direction = "backward")

Start:  AIC=355.68
stroke ~ id + gender + age + hypertension + heart_disease + ever_married + 
    work_type + Residence_type + avg_glucose_level + bmi + smoking_status

                    Df Deviance    AIC
- work_type          1   331.68 353.68
- ever_married       1   331.69 353.69
- gender             1   331.83 353.83
- smoking_status     1   331.95 353.95
- heart_disease      1   332.29 354.29
- avg_glucose_level  1   332.34 354.34
- bmi                1   332.52 354.52
- Residence_type     1   333.24 355.24
<none>                   331.68 355.68
- hypertension       1   334.98 356.98
- id                 1   335.31 357.31
- age                1   391.29 413.29

Step:  AIC=353.68
stroke ~ id + gender + age + hypertension + heart_disease + ever_married + 
    Residence_type + avg_glucose_level + bmi + smoking_status

                    Df Deviance    AIC
- ever_married       1   331.69 351.69
- gender             1   331.83 351.83
- smoking_status     1   331.95 351.95
- heart_d

In [39]:
#생성한 모델과 새롭게 정의한 모델의 summary결과를 확인한다.

summary(m2)


Call:
glm(formula = stroke ~ id + age + hypertension, family = "binomial", 
    data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2166  -0.7522   0.1050   0.7810   2.4496  

Coefficients:
               Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -4.463e+00  5.742e-01  -7.773 7.66e-15 ***
id            1.106e-05  6.115e-06   1.808   0.0706 .  
age           6.891e-02  8.519e-03   8.090 5.98e-16 ***
hypertension  7.083e-01  3.417e-01   2.073   0.0381 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 465.79  on 335  degrees of freedom
Residual deviance: 337.12  on 332  degrees of freedom
AIC: 345.12

Number of Fisher Scoring iterations: 5


In [40]:
#step, update을 생성한 모델을 새롭게 정의할 것. 
#생성한 모델과 새롭게 정의한 모델의 summary결과를 확인한다.

m3 <- update(m, .~ .)
summary(m3)


Call:
glm(formula = stroke ~ id + gender + age + hypertension + heart_disease + 
    ever_married + work_type + Residence_type + avg_glucose_level + 
    bmi + smoking_status, family = "binomial", data = train)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.09644  -0.76500   0.08044   0.77259   2.60667  

Coefficients:
                    Estimate Std. Error z value Pr(>|z|)    
(Intercept)       -4.608e+00  1.376e+00  -3.349 0.000811 ***
id                 1.174e-05  6.208e-06   1.891 0.058568 .  
gender            -1.108e-01  2.828e-01  -0.392 0.695184    
age                6.673e-02  9.827e-03   6.790 1.12e-11 ***
hypertension       6.270e-01  3.509e-01   1.787 0.073987 .  
heart_disease      3.603e-01  4.690e-01   0.768 0.442362    
ever_married      -4.196e-02  4.408e-01  -0.095 0.924175    
work_type          4.412e-03  1.502e-01   0.029 0.976572    
Residence_type    -3.413e-01  2.735e-01  -1.248 0.211963    
avg_glucose_level  2.166e-03  2.668e-03

In [41]:
# Test set을 사용해서 모델을 예측한다. 예측 결과와 원본 데이터와 비교하여 모델의 적합성에 대하여 기술한다. 

#테스트 데이터 확인
test %>% show()

       id gender  age hypertension heart_disease ever_married work_type
1     187      2 20.0            0             0            2         4
9    1261      1 54.0            0             0            1         4
17   2458      2 78.0            0             0            1         4
20   2548      2 81.0            0             0            1         5
36   4655      1 49.0            0             0            1         4
37   4699      1 50.0            0             0            2         2
38   4712      2 81.0            0             1            1         5
46   5714      2 49.0            1             0            1         2
54   7547      1 74.0            0             0            1         4
59   8580      2 77.0            0             0            1         5
66  10552      2 81.0            0             0            1         5
89  13861      2 52.0            1             0            1         5
94  14517      1 56.0            0             0            1   

In [42]:
predict_value <- predict(m, test, type = "response") %>% tibble(predict_value = .)
predict_value %>% show()

# A tibble: 82 × 1
   predict_value
           <dbl>
 1        0.0349
 2        0.348 
 3        0.724 
 4        0.732 
 5        0.326 
 6        0.220 
 7        0.696 
 8        0.366 
 9        0.713 
10        0.658 
# … with 72 more rows


In [43]:
predict_value <- predict(m2, test, type = "response") %>% tibble(predict_value = .)
predict_value %>% show()

# A tibble: 82 × 1
   predict_value
           <dbl>
 1        0.0438
 2        0.326 
 3        0.719 
 4        0.759 
 5        0.262 
 6        0.276 
 7        0.763 
 8        0.422 
 9        0.673 
10        0.719 
# … with 72 more rows


In [44]:
predict_value <- predict(m3, test, type = "response") %>% tibble(predict_value = .)
predict_value %>% show()

# A tibble: 82 × 1
   predict_value
           <dbl>
 1        0.0349
 2        0.348 
 3        0.724 
 4        0.732 
 5        0.326 
 6        0.220 
 7        0.696 
 8        0.366 
 9        0.713 
10        0.658 
# … with 72 more rows


In [45]:
predict_check <- test %>% select(stroke) %>% dplyr::bind_cols(., predict_value) 
predict_check %>% show()

   stroke predict_value
1       0   0.034927288
2       1   0.348211766
3       1   0.724048549
4       1   0.731846142
5       0   0.325983742
6       0   0.219806542
7       1   0.695562633
8       0   0.365976359
9       1   0.713169424
10      1   0.657690097
11      1   0.654606301
12      1   0.656125959
13      0   0.413727871
14      0   0.126051977
15      1   0.851918306
16      0   0.362534063
17      0   0.437956998
18      0   0.371660893
19      1   0.618251833
20      0   0.020303033
21      0   0.154925829
22      1   0.831646436
23      0   0.157157849
24      1   0.685312511
25      0   0.031367901
26      1   0.672378703
27      0   0.009640357
28      1   0.877623745
29      1   0.813968129
30      0   0.144297979
31      0   0.232998743
32      1   0.660231672
33      0   0.196866558
34      1   0.790627582
35      0   0.635059854
36      1   0.165318481
37      0   0.103536588
38      1   0.680248965
39      1   0.386293885
40      1   0.815496033
41      0   0.15

In [46]:
# 예측 결과와 원본 데이터와 비교하여 모델의 적합성에 대하여 기술한다. 
predict(m2, newdata = test, type = "response") %>%
  tibble(predict_stroke = .) %>% bind_cols(test, .)

id    gender age  hypertension heart_disease ever_married work_type
1    187 2      20.0 0            0             2            4        
2   1261 1      54.0 0            0             1            4        
3   2458 2      78.0 0            0             1            4        
4   2548 2      81.0 0            0             1            5        
5   4655 1      49.0 0            0             1            4        
6   4699 1      50.0 0            0             2            2        
7   4712 2      81.0 0            1             1            5        
8   5714 2      49.0 1            0             1            2        
9   7547 1      74.0 0            0             1            4        
10  8580 2      77.0 0            0             1            5        
11 10552 2      81.0 0            0             1            5        
12 13861 2      52.0 1            0             1            5        
13 14517 1      56.0 0            0             1            4        
14 15061 1      40.0 0            0             1            2        
15 17013 1      78.0 1            0             2            4        
16 21381 2      52.0 0            0             1            4        
17 21472 1      52.0 0            1             1            5        
18 22896 2      54.0 0            0             1            4        
19 23410 2      72.0 0            0             1            4        
20 24246 1       7.0 0            0             2            1        
21 24447 2      42.0 0            0             1            4        
22 24669 2      77.0 0            1             1            4        
23 24876 1      35.0 0            0             1            4        
24 26727 2      79.0 0            0             2            4        
25 26893 1       8.0 0            0             2            1        
26 27169 2      66.0 1            0             1            2        
27 27300 2       1.8 0            0             2            1        
28 28291 2      79.0 0            1             1            4        
29 28378 1      61.0 1            1             1            4        
30 28669 2      32.0 0            0             1            4        
⋮  ⋮     ⋮      ⋮    ⋮            ⋮             ⋮            ⋮        
53 45573 2      50   0            0             1            4        
54 46703 1      68   0            1             1            4        
55 47472 2      58   0            0             1            4        
56 50522 2      72   0            0             1            2        
57 51314 2      78   0            0             1            4        
58 51693 2      52   0            0             1            4        
59 52282 1      57   0            0             1            4        
60 53538 2       7   0            0             2            1        
61 56112 1      64   0            1             1            4        
62 56546 1      79   0            1             1            4        
63 56939 2      55   0            0             1            5        
64 58202 2      50   1            0             1            5        
65 58267 1      70   1            0             1            4        
66 58978 2      70   0            1             1            4        
67 59437 2      57   0            0             1            4        
68 60464 1      52   0            0             1            4        
69 60491 2      78   0            0             1            4        
70 60981 2      26   0            0             2            4        
71 64174 2      59   1            0             1            4        
72 65643 2       7   0            0             2            1        
73 66883 2      42   0            0             1            5        
74 67438 2      60   0            0             1            2        
75 68627 1      80   1            1             1            4        
76 69112 1      68   1            1             1            4        
77 69379 2      64   1     

In [47]:
# 예측 결과와 원본 데이터와 비교하여 모델의 적합성에 대하여 기술한다. 
predict(m3, newdata = test, type = "response") %>%
  tibble(predict_stroke = .) %>% bind_cols(test, .)

id    gender age  hypertension heart_disease ever_married work_type
1    187 2      20.0 0            0             2            4        
2   1261 1      54.0 0            0             1            4        
3   2458 2      78.0 0            0             1            4        
4   2548 2      81.0 0            0             1            5        
5   4655 1      49.0 0            0             1            4        
6   4699 1      50.0 0            0             2            2        
7   4712 2      81.0 0            1             1            5        
8   5714 2      49.0 1            0             1            2        
9   7547 1      74.0 0            0             1            4        
10  8580 2      77.0 0            0             1            5        
11 10552 2      81.0 0            0             1            5        
12 13861 2      52.0 1            0             1            5        
13 14517 1      56.0 0            0             1            4        
14 15061 1      40.0 0            0             1            2        
15 17013 1      78.0 1            0             2            4        
16 21381 2      52.0 0            0             1            4        
17 21472 1      52.0 0            1             1            5        
18 22896 2      54.0 0            0             1            4        
19 23410 2      72.0 0            0             1            4        
20 24246 1       7.0 0            0             2            1        
21 24447 2      42.0 0            0             1            4        
22 24669 2      77.0 0            1             1            4        
23 24876 1      35.0 0            0             1            4        
24 26727 2      79.0 0            0             2            4        
25 26893 1       8.0 0            0             2            1        
26 27169 2      66.0 1            0             1            2        
27 27300 2       1.8 0            0             2            1        
28 28291 2      79.0 0            1             1            4        
29 28378 1      61.0 1            1             1            4        
30 28669 2      32.0 0            0             1            4        
⋮  ⋮     ⋮      ⋮    ⋮            ⋮             ⋮            ⋮        
53 45573 2      50   0            0             1            4        
54 46703 1      68   0            1             1            4        
55 47472 2      58   0            0             1            4        
56 50522 2      72   0            0             1            2        
57 51314 2      78   0            0             1            4        
58 51693 2      52   0            0             1            4        
59 52282 1      57   0            0             1            4        
60 53538 2       7   0            0             2            1        
61 56112 1      64   0            1             1            4        
62 56546 1      79   0            1             1            4        
63 56939 2      55   0            0             1            5        
64 58202 2      50   1            0             1            5        
65 58267 1      70   1            0             1            4        
66 58978 2      70   0            1             1            4        
67 59437 2      57   0            0             1            4        
68 60464 1      52   0            0             1            4        
69 60491 2      78   0            0             1            4        
70 60981 2      26   0            0             2            4        
71 64174 2      59   1            0             1            4        
72 65643 2       7   0            0             2            1        
73 66883 2      42   0            0             1            5        
74 67438 2      60   0            0             1            2        
75 68627 1      80   1            1             1            4        
76 69112 1      68   1            1             1            4        
77 69379 2      64   1     